datapath: provide node configuration at runtime (#38244)
cilium-team/cilium: Documentation/contributing/development/datapath_config.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _dpconfig:

Configuring the Datapath
========================

Introduction
~~~~~~~~~~~~

In order for the Cilium datapath to function, it needs access to configuration
data such as feature flags, addresses, timeouts, security IDs and all sorts of
tunables and user configuration. These values are provided by the agent at the

> time of loading the bpf program. This page outlines the configuration mechanism,

Reviewer: s/bpf/BPF/ ? Here and elsewhere

Author: Yep, this was in the back of my mind but forgot, thanks.

some recommendations, how to migrate legacy configuration, as well as practical
examples.

Example

> ~~~~~~~

Reviewer: It feels a bit strange to me to have an “Example” before having any explanation. Maybe change to “Quick tutorial” or “Getting started”?

Author: Ack on the title, perhaps Getting Started makes more sense. I wanted to lead with a practical overview since it's more likely to grab the reader's attention and respect the reader's time. Can't discuss best practices and migration steps before showcasing what the API looks like. I'll add a little paragraph under what's currently "Example" to clarify this, thanks for pointing it out. :+1:

Reviewer: Just to be clear, I think the structure you used is a good idea! Just the naming, “Example”, that felt a bit unintuitive to me, but otherwise no problem. Thanks!

datapath and run ``dpgen`` to generate Go code:

.. code-block:: bash

  make -C bpf -j$(nproc)

This will emit our variable to one of the Go config scaffoldings in the
``pkg/datapath/config`` Go package.

Wiring up Go Values
^^^^^^^^^^^^^^^^^^^

One of the files in package ``config`` will now contain a new struct field that
can be populated at BPF load time.

.. code-block:: go

  type BPFLXC struct {
    ...
    // The endpoint's security ID.
    EndpointID uint16 `config:"endpoint_id"`
    ...
  }

As shown in the preceding snippet, the new struct field carries our helpful
comment we provided in the C code and refers to the ``endpoint_id`` variable we
declared.

.. note::

  At the time of writing, populating Go configuration scaffolding still mostly
  happens in ``pkg/datapath/loader`` and is scattered between a few places. The
  goal is to create StateDB tables for each configuration object. These can be
  managed from Hive Cells and automatically trigger a reload of the necessary
  BPF programs when any of the values change. This document will be updated
  along with these changes.

Now, we need to wire up the field with an actual value. Depending on which

> object you're adding configuration to and depending on whether it's 'node
> configuration' (more below) or object-specific, you may need to look in different

Reviewer: ```suggestion
object you're adding configuration to and depending on whether this configuration is "node
configuration" (more below) or object-specific, you may need to look in different
```

places. For example, adding a value to ``bpf_lxc.c`` like in this example, the
value is typically set in ``endpointRewrites()``:

.. code-block:: go

  func endpointRewrites(...) ... {
    ...
    cfg.InterfaceIfindex = uint32(ep.GetIfIndex())
    ...
  }

.. warning::

  This plumbing needs to be done for every object that needs access to the
  variable! For example, if you declare a variable in a header common to both
  ``bpf_lxc.c`` and ``bpf_host.c``, you'll need to make sure the agent supplies
  the value to both structs.

> If, for some reason, this document no longer matches the codebase, grep around

Reviewer: ```suggestion
If this document no longer matches the codebase, grep around
```

Over time, Hive Cells will be able to write to these structs using StateDB
tables.

Reading the Variable in C
^^^^^^^^^^^^^^^^^^^^^^^^^

We've declared our global config variable. We've generated Go code and wired up
a value from the agent. Now, we need to put the variable to use!

In datapath bpf code, we can refer to it using the ``CONFIG()`` macro. This
macro resolves to a special variable name representing our configuration value,
which could change in the future. The macro is there to avoid cross-cutting code
changes if we ever need to make changes here.

.. note::

  The variable is not a compile-time constant, so it cannot be used to control
  things like BPF map sizes or to initialize other global ``const`` variables at
  compile time.

.. code-block:: C

  CONFIG(endpoint_id)

Use the macro like you would typically use a variable:

.. code-block:: c

  __u16 endpoint_id = CONFIG(endpoint_id);

or in a branch:

.. code-block:: c

  if (CONFIG(endpoint_id) != 0) {
    ...
  }

Node Configuration
~~~~~~~~~~~~~~~~~~

.. warning::

  Historically, most of the agent's configuration was presented to the datapath

>   as 'node configuration' (in ``node_config.h``), but this pattern is

Reviewer: ```suggestion
  as "node configuration" (in ``node_config.h``), but this pattern is
```

>   discouraged going forward and may go away at some point in the future. More on

Reviewer: Are there any CI checks in place to catch new occurrences of “node configuration” changes?

Author: For now, we have all relevant files assigned to `@cilium/loader`, so a small group of in-the-loop folks need to sign off on any changes. There are warning comments littered throughout to alert contributors and direct them to the docs. It's all relatively chaotic code, so I can't think of an effective linter that's easy to implement. For now, we'll direct contributors to the docs if anything shows up in a PR, and we can always lend a hand or make suggestions if needed.

Reviewer: OK, I understand. Thanks!

>   patterns and best practices in a follow-up section.

Reviewer: > in a follow-up section.

Reference and link?
